runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectMetrics.java

```diff
@@ -275,13 +275,40 @@ private boolean matches(MetricsFilter filter, MetricKey key) {
         && matchesScope(key.stepName(), filter.steps());
   }
 
-  private boolean matchesScope(String actualScope, Set<String> scopes) {
+  /**
+  * {@code subPathMatches(actualScope, filterScope)} returns true if {@code filterScope}
+  * represents a path within {@code actualScope}. For example, "foo/bar" is in "a/foo/bar/b",
+  * but not "a/fool/bar/b" or "a/foo/bart/b".
+  */
+  public boolean subPathMatches(String actualScope, String filterScope) {
```

Comment:
Oops. I missed the comments, but your argument makes sense. Set to `haystack` and `needle`.